### PR TITLE
fix(derived_code_mappings): Allow empty or "./" stack trace root

### PR DESCRIPTION
**Pull request by jangjodi**
Fixes WOR-2522

Allows the stack trace root to be empty or ./ depending on if certain cases apply.

### CODE REVIEW

File: src/sentry/integrations/utils/code_mapping.py [hunk 1]
```diff
@@ -276,16 +276,29 @@ def _generate_code_mapping_from_tree(
...
+        if source_path == stacktrace_root:
+            stacktrace_root = ""
+            source_path = ""
+        # Ex. source_path = "static/app", stacktrace_root = "./app"
+        elif "static" in source_path and source_path.replace("static", ".") == stacktrace_root:
+            stacktrace_root = "./"
+            source_path = "static/"
+        # Ex. source_path = "app/", stacktrace_root = "./app"
+        elif "./" in stacktrace_root and stacktrace_root.replace("./", "") == source_path:
+            stacktrace_root = "./"
```
Comment (jangjodi) [inline src/sentry/integrations/utils/code_mapping.py:292]: Changes stack root to empty or `./` and source path to empty or `static/` based on the relevant cases

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 2]
```diff
@@ -41,12 +41,19 @@ def setUp(self):
                     "filename": "./app/utils/handleXhrErrorResponse.tsx",
                     "in_app": True,
                 },
+                {
+                    "filename": "some/path/Test.tsx",
+                    "in_app": True,
```
Comment (jangjodi) [inline tests/sentry/tasks/test_derive_code_mappings.py:46]: Add a new filename for testing if stack root = source path

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 3]
```diff
@@ -41,12 +41,19 @@ def setUp(self):
...
+                },
             ]
         )
 
     def test_find_stacktrace_paths_single_project(self):
         stacktrace_paths = identify_stacktrace_paths(self.event_data)
-        assert stacktrace_paths == ["./app/utils/handleXhrErrorResponse.tsx"]
+        assert set(stacktrace_paths) == {
+            "./app/utils/handleXhrErrorResponse.tsx",
+            "some/path/Test.tsx",
```
Comment (jangjodi) [inline tests/sentry/tasks/test_derive_code_mappings.py:55]: Modified test to include new filename

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 4]
```diff
@@ -81,8 +88,56 @@ def test_derive_code_mappings(self):
             derive_code_mappings(self.project.id, self.event_data)
             code_mapping = RepositoryProjectPathConfig.objects.all()[0]
             # ./app/foo.tsx -> foo.tsx -> static/app/foo.tsx
-            assert code_mapping.stack_root == "./app/"
-            assert code_mapping.source_root == "static/app/"
+            assert code_mapping.stack_root == "./"
+            assert code_mapping.source_root == "static/"
```
Comment (jangjodi) [inline tests/sentry/tasks/test_derive_code_mappings.py:92]: Tests case 2: source_path = "static/app", stacktrace_root = "./app"

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 5]
```diff
@@ -81,8 +88,56 @@ def test_derive_code_mappings(self):
...
+            mock_get_trees_for_org.return_value = {
+                repo_name: RepoTree(
+                    Repo(repo_name, "master"), ["app/utils/handleXhrErrorResponse.tsx"]
+                )
+            }
+            derive_code_mappings(self.project.id, self.event_data)
+            code_mapping = RepositoryProjectPathConfig.objects.all()[0]
+            # ./app/foo.tsx -> foo.tsx -> app/foo.tsx
+            assert code_mapping.stack_root == "./"
+            assert code_mapping.source_root == ""
```
Comment (jangjodi) [inline tests/sentry/tasks/test_derive_code_mappings.py:117]: Tests case 3: source_path = "app/", stacktrace_root = "./app"

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 6]
```diff
@@ -81,8 +88,56 @@ def test_derive_code_mappings(self):
...
+            "sentry.integrations.github.client.GitHubClientMixin.get_trees_for_org"
+        ) as mock_get_trees_for_org:
+            mock_get_trees_for_org.return_value = {
+                repo_name: RepoTree(Repo(repo_name, "master"), ["some/path/Test.tsx"])
+            }
+            derive_code_mappings(self.project.id, self.event_data)
+            code_mapping = RepositoryProjectPathConfig.objects.all()[0]
+            # some/path/Test.tsx -> Test.tsx -> some/path/Test.tsx
+            assert code_mapping.stack_root == ""
+            assert code_mapping.source_root == ""
```
Comment (jangjodi) [inline tests/sentry/tasks/test_derive_code_mappings.py:140]: Tests case 1: source_path = stacktrace_root = "some/"

File: src/sentry/integrations/utils/code_mapping.py [hunk 7]
```diff
@@ -276,16 +276,29 @@ def _generate_code_mapping_from_tree(
             for src_path in repo_tree.files
             if self._potential_match(src_path, frame_filename)
         ]
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:278]:
```suggestion
        ]
        if len(matched_files) != 1
             return []
```

File: src/sentry/integrations/utils/code_mapping.py [hunk 8]
```diff
@@ -276,16 +276,29 @@ def _generate_code_mapping_from_tree(
...
+            source_path = ""
+        # Ex. source_path = "static/app", stacktrace_root = "./app"
+        elif "static" in source_path and source_path.replace("static", ".") == stacktrace_root:
+            stacktrace_root = "./"
+            source_path = "static/"
+        # Ex. source_path = "app/", stacktrace_root = "./app"
+        elif "./" in stacktrace_root and stacktrace_root.replace("./", "") == source_path:
+            stacktrace_root = "./"
+            source_path = ""
         # It is too risky generating code mappings when there's more
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:294]: Replace this whole block with:
```python
return  CodeMapping(
                    repo=repo_tree.repo,
                    stacktrace_root=stacktrace_root,  # sentry/
                    source_path=source_path, # src/sentry/
                )
```

File: src/sentry/integrations/utils/code_mapping.py [hunk 9]
```diff
@@ -276,16 +276,29 @@ def _generate_code_mapping_from_tree(
             for src_path in repo_tree.files
             if self._potential_match(src_path, frame_filename)
         ]
+        # Allow for empty for ./ as stack root if the following cases apply
+        stacktrace_root = f"{frame_filename.root}/"  # sentry
+        source_path = self._get_code_mapping_source_path(matched_files[0], frame_filename)
+        # Ex. source_path = stacktrace_root = "app/"
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:282]:
```suggestion
        # We have a one to one code mapping (e.g. "app/" & "app/")
```

File: src/sentry/integrations/utils/code_mapping.py [hunk 10]
```diff
@@ -276,16 +276,29 @@ def _generate_code_mapping_from_tree(
             for src_path in repo_tree.files
             if self._potential_match(src_path, frame_filename)
         ]
+        # Allow for empty for ./ as stack root if the following cases apply
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:279]:
```suggestion
```

File: src/sentry/integrations/utils/code_mapping.py [hunk 11]
```diff
@@ -276,16 +276,29 @@ def _generate_code_mapping_from_tree(
...
+        stacktrace_root = f"{frame_filename.root}/"  # sentry
+        source_path = self._get_code_mapping_source_path(matched_files[0], frame_filename)
+        # Ex. source_path = stacktrace_root = "app/"
+        if source_path == stacktrace_root:
+            stacktrace_root = ""
+            source_path = ""
+        # Ex. source_path = "static/app", stacktrace_root = "./app"
+        elif "static" in source_path and source_path.replace("static", ".") == stacktrace_root:
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:287]: Try to remove these two last blocks and do this:

```python
elif stacktrace_root.startswith("./"):
    without = stacktrace_root.replace("./", "")
    if source_path == without:
            stacktrace_root = ""
            source_path = ""
    elif source_path.rfind(f"/{without}"):
            stacktrace_root = ""
            source_path = source_path.replace(f"/{without}", "/")
```

File: src/sentry/integrations/utils/code_mapping.py [hunk 12]
```diff
@@ -276,16 +276,29 @@ def _generate_code_mapping_from_tree(
             for src_path in repo_tree.files
             if self._potential_match(src_path, frame_filename)
         ]
+        # Allow for empty for ./ as stack root if the following cases apply
+        stacktrace_root = f"{frame_filename.root}/"  # sentry
```
Comment (armenzg) [inline src/sentry/integrations/utils/code_mapping.py:280]: Move all of this code up to the return into a function called `_normalized_stack_and_source_roots`. You can then write unit tests for it.

```python
stacktrace_root, source_path = _normalized_stack_and_source_roots(f"{frame_filename.root}/", self._get_code_mapping_source_path(matched_files[0], frame_filename))
```

File: src/sentry/integrations/utils/code_mapping.py [hunk 13]
```diff
@@ -266,6 +266,24 @@ def _get_code_mapping_source_path(self, src_file: str, frame_filename: FrameFile
...
+        # stacktrace_root starts with "./"
+        elif stacktrace_root.startswith("./"):
+            without = stacktrace_root.replace("./", "")
+            if source_path == without:
+                stacktrace_root = "./"
+                source_path = ""
+            elif source_path.rfind(f"/{without}"):
+                stacktrace_root = "./"
+                source_path = source_path.replace(f"/{without}", "/")
+        return (stacktrace_root, source_path)
```
Comment (jangjodi) [inline src/sentry/integrations/utils/code_mapping.py:285]: Function to allow for empty or "./" as stack and source roots

File: src/sentry/integrations/utils/code_mapping.py (hunk 7 above)
Comment (jangjodi) [reply src/sentry/integrations/utils/code_mapping.py:296]: This was just missing a `:` but it was added

File: tests/sentry/integrations/utils/test_code_mapping.py [hunk 14]
```diff
@@ -262,3 +262,24 @@ def test_list_file_matches_multiple(self):
             },
         ]
         assert matches == expected_matches
+
+    def test_normalized_stack_and_source_roots_static(self):
+        stacktrace_root, source_path = self.code_mapping_helper._normalized_stack_and_source_roots(
+            "./app/", "static/app/"
+        )
```
Comment (jangjodi) [inline tests/sentry/integrations/utils/test_code_mapping.py:269]: Unit test for `_normalized_stack_and_source_roots` function

The following tests check for different cases.

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 15]
```diff
@@ -81,8 +88,56 @@ def test_derive_code_mappings(self):
             derive_code_mappings(self.project.id, self.event_data)
             code_mapping = RepositoryProjectPathConfig.objects.all()[0]
             # ./app/foo.tsx -> foo.tsx -> static/app/foo.tsx
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:90]:
```suggestion
            # ./app/foo.tsx -> app/foo.tsx -> static/app/foo.tsx
```

The stack root drops "./", thus, `app/foo.tsx`

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 16]
```diff
@@ -81,8 +88,56 @@ def test_derive_code_mappings(self):
...
+            "sentry.integrations.github.client.GitHubClientMixin.get_trees_for_org"
+        ) as mock_get_trees_for_org:
+            mock_get_trees_for_org.return_value = {
+                repo_name: RepoTree(
+                    Repo(repo_name, "master"), ["app/utils/handleXhrErrorResponse.tsx"]
+                )
+            }
+            derive_code_mappings(self.project.id, self.event_data)
+            code_mapping = RepositoryProjectPathConfig.objects.all()[0]
+            # ./app/foo.tsx -> foo.tsx -> app/foo.tsx
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:115]:
```suggestion
            # ./app/foo.tsx -> app/foo.tsx -> app/foo.tsx
```
The stack root drops "./", thus, `app/foo.tsx`

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 17]
```diff
@@ -81,8 +88,56 @@ def test_derive_code_mappings(self):
...
+            assert code_mapping.stack_root == "./"
+            assert code_mapping.source_root == ""
+            assert code_mapping.repository.name == repo_name
+
+    @responses.activate
+    @with_feature("organizations:derive-code-mappings")
+    def test_derive_code_mappings_empty(self):
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:122]:
```suggestion
    def test_derive_code_mappings_one_to_one_match(self):
```

File: tests/sentry/tasks/test_derive_code_mappings.py [hunk 18]
```diff
@@ -81,8 +88,56 @@ def test_derive_code_mappings(self):
...
+    @responses.activate
+    @with_feature("organizations:derive-code-mappings")
+    def test_derive_code_mappings_not_static(self):
+        repo_name = "foo/bar"
+        self.create_integration(
+            organization=self.organization,
+            provider="github",
+            external_id=self.organization.id,
+            metadata={"domain_name": "github.com/Test-Org"},
+        )
```
Comment (armenzg) [inline tests/sentry/tasks/test_derive_code_mappings.py:104]: Ideally, this block of code would be part of the `setUp` to prevent repetition.

**Review by armenzg — COMMENTED** [review]
Good coverage score :)

image

**Review by armenzg — APPROVED** [review]

**Pull request merged**